presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java

```diff
@@ -546,19 +550,37 @@ protected Object visitCoalesceExpression(CoalesceExpression node, Object context
             List<Object> values = node.getOperands().stream()
                     .map(value -> processWithExceptionHandling(value, context))
                     .filter(Objects::nonNull)
+                    .flatMap(expression -> {
+                        if (expression instanceof CoalesceExpression) {
+                            return ((CoalesceExpression) expression).getOperands().stream();
+                        }
+                        return Stream.of(expression);
+                    })
                     .collect(Collectors.toList());
 
             if ((!values.isEmpty() && !(values.get(0) instanceof Expression)) || values.size() == 1) {
                 return values.get(0);
             }
-
-            List<Expression> expressions = values.stream()
-                    .map(value -> toExpression(value, type))
-                    .collect(Collectors.toList());
+            ImmutableList.Builder<Expression> operandsBuilder = ImmutableList.builder();
+            Set<Expression> visitedExpression = new HashSet<>();
+            for (Object value : values) {
+                Expression expression = toExpression(value, type);
+                if (!isDeterministic(expression) || visitedExpression.add(expression)) {
+                    operandsBuilder.add(expression);
+                }
+                if (!canEvaluateToNull(expression)) {
```

Comment:
This should actually evaluate the expression to determine if the value is a constant non-null. Of course, it needs to guard against exceptions and bail out if that happens to preserve lazy evaluation semantics.

Reply:
Actual evaluation is being performed at  `processWithExceptionHandling(value, context)` 
`canEvaluateToNull` currently return false if the `Expression` is of Type `Literal` (except for `NullLiteral`). In future we can evaluate based on the input from the connector if it could have some null values or if it passes a comparison filter criteria (like a > 10, now we can assume that a can't be null)

Reply:
> This should actually evaluate the expression to determine if the value is a constant non-null. Of course, it needs to guard against exceptions and bail out if that happens to preserve lazy evaluation semantics.

How you could evaluate it? It can accept multiple inputs, e.g: `i1, i2, i3, i4, ..., in` and for only some of the combinations it could return null.

Reply:
> This should actually evaluate the expression to determine if the value is a constant non-null

I guess one could evaluate expression to see if it's constant and if it is, then check if it is non-null.

In practice, whenever expression has symbol references, it will be nullable.

Reply:
> In practice, whenever expression has symbol references, it will be nullable.

Ideally yes but we can make use of `NullabilityAnalyzer` or `NullityEvaluator` to fetch even more info about whether this `SymbolReference` evaluates to null. Like in `ColumnHandle` I could provide a extra information that will say whether this Column has only non null values so now I could consider that when I am trying to simplify any `Expression`

Reply:
> Like in ColumnHandle I could provide a extra information that will say whether this Column has only non null values so now I could consider that when I am trying to simplify any Expression

We could and should, but it seems to be outside of the scope of this PR. Subsequent improvements are welcome though.

Nullability inference is a tricky problem and comes out in various places like equality inference and operator (e.g: semi-join) semantics.

Reply:
> Nullability inference is a tricky problem and comes out in various places like equality inference and operator (e.g: semi-join) semantics.

Yes , but instead of using `node instanceof Literal && !(node instanceofNull)` I tried using a evaluator which we can be improvised.

Reply:
Since the engine is not currently able to reason about nullability in general, let's get rid of that and do the check against literal null. At the point in the future where we add that ability we can introduce it here.